packages/eslint-plugin/src/rules/no-unused-private-class-members.ts

```diff
@@ -0,0 +1,261 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import { createRule } from '../util';
+
+type Options = [];
+export type MessageIds = 'unusedPrivateClassMember';
+
+interface PrivateMember {
+  declaredNode:
+    | TSESTree.MethodDefinitionComputedName
+    | TSESTree.MethodDefinitionNonComputedName
+    | TSESTree.PropertyDefinitionComputedName
+    | TSESTree.PropertyDefinitionNonComputedName;
+  isAccessor: boolean;
+}
+
+export default createRule<Options, MessageIds>({
+  name: 'no-unused-private-class-members',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow unused private class members',
+      extendsBaseRule: true,
+      requiresTypeChecking: false,
+    },
+
+    schema: [],
+
+    messages: {
+      unusedPrivateClassMember:
+        "'{{classMemberName}}' is defined but never used.",
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const trackedClassMembers: Map<string, PrivateMember>[] = [];
+
+    /**
+     * The core ESLint rule tracks methods by adding an extra property of
+     * "isUsed" to the method, which obviously violates type safety. Instead, in
```

Comment:
> which obviously violates type safety

[Docs] _"obvious"_ is a strong word, and IMO something to be avoided. Not everything is obvious to everyone. If something is described by an adjective like _"easy"_ or _"obvious"_, then a reader doesn't get it at first, that's not a great feeling on their end.

```suggestion
     * "isUsed" to the method. Instead of modifying the AST, in
```

Aside: IMO the `.isUsed` thing is really more of a "bug" in ESLint core. They shouldn't be modifying the AST in arbitrary ways like that. If nobody else files an issue by the time this PR is resolved, I'll try to find time to. 🙂